CoreFoundation/NumberDate.subproj/CFTimeZone.c

```diff
@@ -183,6 +185,121 @@ static void __CFTimeZoneGetOffset(CFStringRef timezone, int32_t *offset) {
 
     RegCloseKey(hKey);
 }
+#elif TARGET_OS_ANDROID
+typedef Boolean (*__CFAndroidTimeZoneListEnumerateCallback)(char name[40], int32_t start, int32_t length, FILE *fp, void *context);
+static void __CFAndroidTimeZoneListEnumerate(__CFAndroidTimeZoneListEnumerateCallback callback, void *context) {
+    // The best reference should be Android Bionic's libc/tzcode/bionic.cpp
+    static const char *tzDataFiles[] = {
+        "/data/misc/zoneinfo/current/tzdata",
+        "/system/usr/share/zoneinfo/tzdata"
+    };
+
+    Boolean done = FALSE;
+    for (int idx = 0; idx < sizeof(tzDataFiles)/sizeof(tzDataFiles[0]) && !done; idx++) {
+        FILE *fp = fopen(tzDataFiles[idx], "rb");
+        if (!fp) {
+            continue;
+        }
+
+        char header[24];
+        if (fread(header, 1, sizeof(header), fp) != sizeof(header)) {
+            fclose(fp);
+            continue;
+        }
+        if (strncmp(header, "tzdata", 6) != 0) {
+            fclose(fp);
+            continue;
+        }
+        int32_t indexOffset;
+        memcpy(&indexOffset, &header[12], sizeof(int32_t));
+        indexOffset = betoh32(indexOffset);
+        int32_t dataOffset;
+        memcpy(&dataOffset, &header[16], sizeof(int32_t));
+        dataOffset = betoh32(dataOffset);
+        if (indexOffset < 0 || dataOffset < indexOffset) {
+            fclose(fp);
+            continue;
+        }
+        if (fseek(fp, indexOffset, SEEK_SET) != 0) {
+            fclose(fp);
+            continue;
+        }
+
+        char entry[52];
+        size_t indexSize = dataOffset - indexOffset;
+        size_t zoneCount = indexSize / sizeof(entry);
+        if (zoneCount * sizeof(entry) != indexSize) {
+            fclose(fp);
+            continue;
+        }
+        for (size_t idx = 0; idx < zoneCount; idx++) {
+            if (fread(entry, 1, sizeof(entry), fp) != sizeof(entry)) {
+                break;
+            }
+            int32_t start;
+            memcpy(&start, &entry[40], sizeof(int32_t));
+            start = betoh32(start);
+            start += dataOffset;
+            int32_t length;
+            memcpy(&length, &entry[44], sizeof(int32_t));
+            length = betoh32(length);
+            if (start < 0 || length < 0) {
+                break;
+            }
+
+            done = callback(entry, start, length, fp, context);
+            if (done) {
+                break;
+            }
+        }
+
+        fclose(fp);
+    }
+}
+
+static Boolean __CFCopyAndroidTimeZoneListCallback(char name[40], int32_t start, int32_t length, FILE *fp, void *context) {
+    CFMutableArrayRef result = (CFMutableArrayRef)context;
+    size_t lenght = strnlen(name, 40);
+    CFStringRef timeZoneName = CFStringCreateWithCString(kCFAllocatorSystemDefault, name, kCFStringEncodingASCII);
+    CFArrayAppendValue(result, timeZoneName);
+    CFRelease(timeZoneName);
+    return FALSE;
+}
+
+struct __CFTimeZoneDataCreateContext {
```

Comment:
Why the `struct` shenanigans?  The callback is entirely local, and the caller is local as well.  You could just adjust the signature to pass the `tzname` and `tzdata`.

Reply:
When I normally pass function pointers and they need to share data, there's only one `context` pointer, and you have to figure out how to pack all the needed data into one pointer.

For enumerating the known time zones, only one pointer is needed (the `CFMutableArrayRef`), but for finding out about one time zone, I need both the required time zone identifier and a double pointer to where to leave the data.

I can indeed pass many more pointers, but this was a more general solution, and a pattern everyone will recognize.

Reply:
I think that both patterns are ones that everyone will recognise.  It is very common in C to have custom callbacks with an arbitrary set of parameters.